### PR TITLE
[AIRFLOW-4741] Include Sentry for Airflow

### PR DESCRIPTION
**Pull request by tiopi**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-XXX
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
This PR is to add Sentry SDK functionality to Airflow. The functionality includes adding tags per task event and breadcrumbs of previous tasks that ran on the DAG. 

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

File: airflow/contrib/hooks/sentry_hook.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        integrations = [sentry_celery]
+        ignore_logger("airflow.task")
+
+        self.conn_id = self.get_connection(sentry_conn_id)
+        self.dsn = self.conn_id.host
+        init(dsn=self.dsn, integrations=integrations)
+
+        if not getattr(TaskInstance, "_sentry_integration_", False):
+            TaskInstance.__init__ = add_sentry
+            TaskInstance.clear_xcom_data = new_clear_xcom
```
Comment (ashb) [inline airflow/contrib/hooks/sentry_hook.py]: Instead of hooking on `clear_xcom_data` please use the same trick against the `pre_execute` fn.

File: airflow/contrib/hooks/sentry_hook.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            TaskInstance.ds = ds
```
Comment (ashb) [inline airflow/contrib/hooks/sentry_hook.py]: Why is this needed? Where is it used for?

File: airflow/contrib/hooks/sentry_hook.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
+from airflow.hooks.base_hook import BaseHook
+from airflow.utils.db import provide_session
+from airflow.models import TaskInstance
+
+from sentry_sdk.integrations.celery import CeleryIntegration
```
Comment (ashb) [inline airflow/contrib/hooks/sentry_hook.py]: Celery is optional in Airflow - is it worth including this only if Airflow is configured to use it?

File: airflow/contrib/hooks/sentry_hook.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+from sentry_sdk.integrations.logging import ignore_logger
+from sentry_sdk import configure_scope, add_breadcrumb, init
+
+original_task_init = TaskInstance.__init__
+original_clear_xcom = TaskInstance.clear_xcom_data
+SCOPE_TAGS = frozenset(("task_id", "dag_id", "execution_date", "ds", "operator"))
```
Comment (ashb) [inline airflow/contrib/hooks/sentry_hook.py]: Why execution_date  and ds? Why operator too?

File: airflow/contrib/hooks/sentry_hook.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+
+@provide_session
+def new_clear_xcom(self, session=None):
+    """
+    Add breadcrumbs just before task is executed.
+    """
+    for task in self.task.get_flat_relatives(upstream=True):
+        state = get_task_instance_attr(self, task.task_id, "state")
+        operation = get_task_instance_attr(self, task.task_id, "operator")
```
Comment (ashb) [inline airflow/contrib/hooks/sentry_hook.py]: No need to go to the DB for this one
```suggestion
        operation = task.__class__.__name__
```

File: airflow/contrib/hooks/sentry_hook.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        sentry_celery = CeleryIntegration()
+        integrations = [sentry_celery]
+        ignore_logger("airflow.task")
+
+        self.conn_id = self.get_connection(sentry_conn_id)
+        self.dsn = self.conn_id.host
+        init(dsn=self.dsn, integrations=integrations)
+
+        if not getattr(TaskInstance, "_sentry_integration_", False):
+            TaskInstance.__init__ = add_sentry
```
Comment (ashb) [inline airflow/contrib/hooks/sentry_hook.py]: Doing this doesn't feel very clean. Is there another way?

Where is this function actually called from too?

File: airflow/www/static/connection_form.js [hunk 7]
```diff
@@ -0,0 +1,115 @@
+/**
```
Comment (ashb) [inline airflow/www/static/connection_form.js]: I don't think you meant to add this file?

File: airflow/www/static/js/connection_form.js [hunk 8]
```diff
@@ -62,6 +62,12 @@ $(document).ready(function () {
         'login': 'Username',
       }
     },
+    sentry: {
+      hidden_fields:['host', 'schema', 'login', 'password', 'port', 'extra'],
+      relabeling: {
+        'conn_id': 'dsn'
```
Comment (ashb) [inline airflow/www/static/js/connection_form.js]: That doesn't seem right - conn_id is a name to lookup the connection by (i.e. `sentry_default` or similar).

Sentry has an API key right? If so that should be stored in the password field, or the entire dsn should be in the `host` field.

**Review by ashb — CHANGES_REQUESTED** [review]
Changes needed to the code as it stands, but I'm not sure I like the money-patching approach.

Does this PR even _work_? Nothing is loading the SentryHook to add the integrations.

Could it instead be attached to the existing task success/failure callbacks?

File: airflow/www/static/js/connection_form.js (hunk 8 above)
Comment (tiopi) [reply airflow/www/static/js/connection_form.js]: You are right, the previous file was a mistake, and I meant to place the dsn in the host field.

File: airflow/contrib/hooks/sentry_hook.py (hunk 3 above)
Comment (tiopi) [reply airflow/contrib/hooks/sentry_hook.py]: Good point!

File: airflow/contrib/hooks/sentry_hook.py (hunk 2 above)
Comment (tiopi) [reply airflow/contrib/hooks/sentry_hook.py]: Removed.

File: airflow/contrib/hooks/sentry_hook.py (hunk 4 above)
Comment (tiopi) [reply airflow/contrib/hooks/sentry_hook.py]: I removed ds, so it will only be execution_date and operator. These are to keep tags of tasks that fail. These tags are helpful to group and know whether the failure is due to an operator or perhaps just the execution date.

File: airflow/contrib/hooks/sentry_hook.py (hunk 6 above)
Comment (tiopi) [reply airflow/contrib/hooks/sentry_hook.py]: This function patches the init function of a TaskInstance. I went for the monkey patching route since I believe that someone who is using airflow does not need any of the sentry code. However, if they do opt in by adding the DSN to the connections, then the code is added to the init function of the tasks.

Comment (mik-laj) [reply airflow/contrib/hooks/sentry_hook.py]: I am very glad that this integration appears. I have doubts whether the current implementation is good. I think the init logic of this code does not necessarily have to be in the hook. I think that it is not a problem to initiate this integration in the core.

I also think that hooking methods in this way is not a good idea. This can be very confusing to other people.
